### PR TITLE
GUI & Graphics Testing

### PR DESCRIPTION
**Pull request by judicaelclair**
Rady:
- `TrackedSet` implemented.
- Testing `Tracked` structures.
- Camera control & screenshots.
- Basic PyAutoGUI example for testing graphics.
- Polished UI, changed to aspect2d & added resize buttons.

Judi:
- Cleaned up GUI code.
- Fixed bugs related to using aspect2d, namely dragging was going bonkers before.
- Saving `SimulatorConfig` state with `SimulatorConfigState` and ability to dump arbitrary objects to json.
- Added `ReinitEvent` and changed the meaning of `ResetEvent` to just mean resetting the current algorithm, not a full re-initialisation.
- Resetting algorithm with simulator config window reset button (rendering gets buggy but I need changes from other branches to fix).
- GUI cleanup
- Ability to change agent and goal position.

### CODE REVIEW

File: src/simulator/views/map/map_view.py [hunk 1]
```diff
@@ -294,20 +305,26 @@ def HDScreenShot(self):
...
-        x,y,z = self.__world.getPos()
-        npCam.setPos(x,y+0.5,z+77.3)
-        npCam.setP(-90)
+        x, y, z = self.__world.getPos()
+        # 3d case
+        if c > 1:
+            npCam.setPos(x-a * 2.1, y - max_height * 3.6, z)
+            npCam.setH(-30)
+        else:
+            npCam.setPos(x, y, z + max_width * 2 + c)
```
Comment (Nabushika) [inline src/simulator/views/map/map_view.py:314]: +c seems a little extraneous here, if we know it's zero?

**Review by Nabushika — APPROVED** [review]
seems strange that tracked container doesn't keep track of the added and removed elements separately, but if it works for you...

**Comment by judicaelclair** [comment]
> seems strange that tracked container doesn't keep track of the added and removed elements separately, but if it works for you...

We could in the future track both, but it doesn't help for the current optimisations.

File: src/simulator/views/map/map_view.py (hunk 1 above)
Comment (judicaelclair) [reply src/simulator/views/map/map_view.py:314]: `c` is 1 in this case, but yh maybe best to remove it. @rm1819

Comment (judicaelclair) [reply src/simulator/views/map/map_view.py:314]: Cleaning this up now

Comment (judicaelclair) [reply src/simulator/views/map/map_view.py]: Done

**Review by JohnYao93 — APPROVED** [review]
all looks good

**Pull request merged**